akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedBehavior.scala

```diff
@@ -0,0 +1,113 @@
+      val writerUuid: String = UUID.randomUUID.toString
+      WriterIdentity(instanceId, writerUuid)
+    }
+  }
+  @InternalApi private[akka] final case class WriterIdentity(instanceId: Int, writerUuid: String)
+
+  /** INTERNAL API: Protocol used internally by the eventsourced behaviors, ever exposed to user-land */
+  @InternalApi private[akka] sealed trait EventsourcedProtocol
+  @InternalApi private[akka] case object RecoveryPermitGranted extends EventsourcedProtocol
+  @InternalApi private[akka] final case class JournalResponse(msg: akka.persistence.JournalProtocol.Response) extends EventsourcedProtocol
+  @InternalApi private[akka] final case class SnapshotterResponse(msg: akka.persistence.SnapshotProtocol.Response) extends EventsourcedProtocol
+  @InternalApi private[akka] final case class RecoveryTickEvent(snapshot: Boolean) extends EventsourcedProtocol
+  @InternalApi private[akka] final case class ReceiveTimeout(timeout: akka.actor.ReceiveTimeout) extends EventsourcedProtocol
+
+  /** INTERNAL DSL */
+  private[akka] implicit class UnlessStopped[T](val b: Behavior[T]) extends AnyVal {
+    /** Returns `other` behavior if the wrapped one is `Stopped`. */
+    def unlessStopped(other: ⇒ Behavior[T]): Behavior[T] =
+      b match {
+        case _: StoppedBehavior[T @unchecked] ⇒ b // still the stopped one
+        case _                                ⇒ other
+      }
+  }
+
+  implicit object PersistentBehaviorLogSource extends LogSource[EventsourcedBehavior[_, _, _]] {
+    override def genString(b: EventsourcedBehavior[_, _, _]): String = {
+      val behaviorShortName = b match {
+        case _: EventsourcedRunning[_, _, _]                  ⇒ "running"
+        case _: EventsourcedRecoveringEvents[_, _, _]         ⇒ "recover-evts"
+        case _: EventsourcedRecoveringSnapshot[_, _, _]       ⇒ "recover-snap"
+        case _: EventsourcedRequestingRecoveryPermit[_, _, _] ⇒ "awaiting-permit"
+      }
+      s"PersistentBehavior[id:${b.persistenceId}][${b.context.self.path}][$behaviorShortName]"
+    }
+  }
+
+}
+
+trait EventsourcedBehavior[Command, Event, State] {
+  import EventsourcedBehavior._
+  import akka.actor.typed.scaladsl.adapter._
+
+  def context: ActorContext[Any]
+  def timers: TimerScheduler[Any]
+
+  type C = Command
+  type AC = ActorContext[C]
+  type E = Event
+  type S = State
+
+  // used for signaling intent in type signatures
+  type SeqNr = Long
+
+  // FIXME make this not even visible in the state that is "running", we'll never call it there again
+  def persistenceId: String
+
+  def initialState: State
+  def commandHandler: PersistentBehaviors.CommandHandler[Command, Event, State]
+  def eventHandler: (State, Event) ⇒ State
+
+  // FIXME make this not even visible in the state that is "running", we'll never call it there again
+  def recoveryCompleted: (ActorContext[Command], State) ⇒ Unit
+  def snapshotWhen: (State, Event, SeqNr) ⇒ Boolean
```

Comment:
I did a small rename snapshotOn => snapshotWhen which sounds more like our operators in streams etc

Reply:
:+1: